File: libs/ultra-infer/python/ultra_infer/vision/perception/paddle3d/petr.py
# Copyright (c) 2024 PaddlePaddle Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import absolute_import

from .... import ModelFormat, UltraInferModel
from .... import c_lib_wrap as C


class PetrPreprocessor:
    def __init__(self, config_file):
        """Create a preprocessor for Petr"""
        self._preprocessor = C.vision.perception.PetrPreprocessor(config_file)

    def run(self, input_ims):
        """Preprocess input images for Petr

        :param: input_ims: (list of numpy.ndarray)The input image
        :return: list of FDTensor
        """
        return self._preprocessor.run(input_ims)


class PetrPostprocessor:
    def __init__(self):
        """Create a postprocessor for Petr"""
        self._postprocessor = C.vision.perception.PetrPostprocessor()

    def run(self, runtime_results):
        """Postprocess the runtime results for Petr

        :param: runtime_results: (list of FDTensor)The output FDTensor results from runtime
        :return: list of PerceptionResult(If the runtime_results is predict by batched samples, the length of this list equals to the batch size)
        """
        return self._postprocessor.run(runtime_results)


class Petr(UltraInferModel):
    def __init__(
        self,
        model_file,
        params_file,
        config_file,
        runtime_option=None,
        model_format=ModelFormat.PADDLE,
    ):
        """Load a SMoke model exported by Petr.

        :param model_file: (str)Path of model file, e.g ./petr.pdmodel
        :param params_file: (str)Path of parameters file, e.g ./petr.pdiparams
        :param config_file: (str)Path of config file, e.g ./infer_cfg.yaml
        :param runtime_option: (ultra_infer.RuntimeOption)RuntimeOption for inference this model, if it's None, will use the default backend on CPU
        :param model_format: (ultra_infer.ModelForamt)Model format of the loaded model
        """
        super(Petr, self).__init__(runtime_option)

        self._model = C.vision.perception.Petr(
            model_file, params_file, config_file, self._runtime_option, model_format
        )
        assert self.initialized, "Petr initialize failed."

    def predict(self, input_image):
        """Detect an input image

        :param input_image: (numpy.ndarray)The input image data, 3-D array with layout HWC, BGR format
        :param conf_threshold: confidence threshold for postprocessing, default is 0.25
        :param nms_iou_threshold: iou threshold for NMS, default is 0.5
        :return: PerceptionResult
        """
        return self._model.predict(input_image)

    def batch_predict(self, images):
        """Classify a batch of input image

        :param im: (list of numpy.ndarray) The input image list, each element is a 3-D array with layout HWC, BGR format
        :return list of PerceptionResult
        """

        return self._model.batch_predict(images)

    @property
    def preprocessor(self):
        """Get PetrPreprocessor object of the loaded model

        :return PetrPreprocessor
        """
        return self._model.preprocessor

    @property
    def postprocessor(self):
        """Get PetrPostprocessor object of the loaded model

        :return PetrPostprocessor
        """
        return self._model.postprocessor

File: deploy/ultra-infer/ultra_infer/vision/perception/paddle3d/petr/petr.cc
// Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "ultra_infer/vision/perception/paddle3d/petr/petr.h"

namespace ultra_infer {
namespace vision {
namespace perception {

Petr::Petr(const std::string &model_file, const std::string &params_file,
           const std::string &config_file, const RuntimeOption &custom_option,
           const ModelFormat &model_format)
    : preprocessor_(config_file) {
  valid_cpu_backends = {Backend::PDINFER};
  valid_gpu_backends = {Backend::PDINFER};

  runtime_option = custom_option;
  runtime_option.model_format = model_format;
  runtime_option.model_file = model_file;
  runtime_option.params_file = params_file;
  runtime_option.paddle_infer_option.enable_mkldnn = false;
  initialized = Initialize();
}

bool Petr::Initialize() {
  if (!InitRuntime()) {
    FDERROR << "Failed to initialize ultra_infer backend." << std::endl;
    return false;
  }
  return true;
}

bool Petr::Predict(const cv::Mat &images, PerceptionResult *results) {
  FDERROR << "Petr inference only support 6(V1) or 12(V2) images" << std::endl;
  return false;
}

bool Petr::BatchPredict(const std::vector<cv::Mat> &images,
                        std::vector<PerceptionResult> *results) {
  if ((images.size() != 6) && (images.size() != 12)) {
    FDERROR << "Petr only support 6(V1) or 12(V2) images";
    return false;
  }
  std::vector<FDMat> fd_images = WrapMat(images);

  if (!preprocessor_.Run(&fd_images, &reused_input_tensors_)) {
    FDERROR << "Failed to preprocess the input image." << std::endl;
    return false;
  }

  // Note: un-commented the codes below to show the debug info.
  // reused_input_tensors_[0].PrintInfo();
  // reused_input_tensors_[1].PrintInfo();
  // reused_input_tensors_[2].PrintInfo();

  reused_input_tensors_[0].name = InputInfoOfRuntime(0).name;
  reused_input_tensors_[1].name = InputInfoOfRuntime(1).name;
  if (images.size() == 12) {
    // for Petr V2 timestamp
    reused_input_tensors_[2].name = InputInfoOfRuntime(2).name;
  } else {
    // for Petr V1
    reused_input_tensors_.pop_back();
  }

  if (!Infer(reused_input_tensors_, &reused_output_tensors_)) {
    FDERROR << "Failed to inference by runtime." << std::endl;
    return false;
  }

  if (!postprocessor_.Run(reused_output_tensors_, results)) {
    FDERROR << "Failed to postprocess the inference results by runtime."
            << std::endl;
    return false;
  }
  return true;
}

} // namespace perception
} // namespace vision
} // namespace ultra_infer
